### PR TITLE
Fix #2041: Add feature for slack notification

### LINKED ISSUE #2041: Add feature for slack notification
## Current Behaviour

Currently, when a challenge is created, submission is made or a contact message is received, EvalAI Admin doesn't get any notification.

## Expected Behaviour

A slack notification should be added for all the three events.

## Deliverables
- [ ] Create a common function in `base/utils` which will take input a slack webhook URL and a message to be sent to the slack.
- [ ] Add the notification for following events:
	- [ ] When a challenge zip is uploaded
	- [ ] When a contact message is received
	- [ ] When a submission is made
- [ ] Write test cases.

## Starter Code
```
def send_slack_notification(webhook=settings.SLACK_WEBHOOKS['default'], message=""):
    '''Send slack notification to any workspace
    
    Keyword Arguments:
        webhook {string} -- slack webhook URL (default: {settings.SLACK_WEBHOOKS['default']})
        message {str} -- JSON/Text message to be sent to slack (default: {""})
    '''

    try:
        response = requests.post(
            webhook,
            data=json.dumps({'text': str(message)}),
            headers={'Content-Type': 'application/json'}
        )
    except Exception as e:
        logger.info('Exception raised while sending slack notification. \n Exception message: {}'.format(e))
```

Please feel free to ask if you have any doubts!

### PR DESCRIPTION
**Pull request by Sanji515**
#### Changes proposed in this pull request:

- Resolved conflicts, tested the functionality using real `web hook url`.

- Enhanced the message body to look more better on the slack.

- Updated the requirements with the `responses>=0.9.0` to support the callbacks.

- Removed the slack notification when making new submission.

- Added the conditions to work well in `dev` environment without API failing.

- Updated tests

Thanks @yashdusing!

### CODE REVIEW

File: apps/challenges/views.py [hunk 1]
```diff
@@ -1123,6 +1124,17 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
             zip_config.challenge = challenge
             zip_config.save()
+
+            if settings.SLACK_WEB_HOOK_URL != "http://testslackwebhook.com/webhook":
+                slack_challenge_details = {
+                    'challenge_title': challenge.title,
+                    'zip_file_link': CHALLENGE_ZIP_DOWNLOAD_LOCATION
+                }
+                send_slack_notification(
+                    message="A *new challenge* has been uploaded to EvalAI. \n *User email*: {}\n *Challenge details*: {}"
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: `A *new challenge* has been uploaded to EvalAI. \n *User email*: {}\n *Challenge details*: {}`
`A *challenge* config is uploaded from {}. \n The challenge title is: {}`

File: apps/challenges/views.py [hunk 2]
```diff
@@ -1123,6 +1124,17 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
 
             zip_config.challenge = challenge
             zip_config.save()
+
+            if settings.SLACK_WEB_HOOK_URL != "http://testslackwebhook.com/webhook":
+                slack_challenge_details = {
+                    'challenge_title': challenge.title,
+                    'zip_file_link': CHALLENGE_ZIP_DOWNLOAD_LOCATION
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: Please remove this.

File: apps/challenges/views.py [hunk 3]
```diff
@@ -1123,6 +1124,17 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
 
             zip_config.challenge = challenge
             zip_config.save()
+
+            if settings.SLACK_WEB_HOOK_URL != "http://testslackwebhook.com/webhook":
+                slack_challenge_details = {
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: we can remove it as well.

File: apps/challenges/views.py [hunk 4]
```diff
@@ -1123,6 +1124,17 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
+                    'challenge_title': challenge.title,
+                    'zip_file_link': CHALLENGE_ZIP_DOWNLOAD_LOCATION
+                }
+                send_slack_notification(
+                    message="A *new challenge* has been uploaded to EvalAI. \n *User email*: {}\n *Challenge details*: {}"
+                    .format(request.user.email, slack_challenge_details)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: use `challenge.title`

File: apps/web/views.py [hunk 5]
```diff
@@ -110,6 +109,12 @@ def contact_us(request):
             response_data = {
                 "message": "We have received your request and will contact you shortly."
             }
+
+            if settings.SLACK_WEB_HOOK_URL != "http://testslackwebhook.com/webhook":
```
Comment (RishabhJain2018) [inline apps/web/views.py]: Please use `if not settings.DEBUG`

File: apps/web/views.py [hunk 6]
```diff
@@ -110,6 +109,12 @@ def contact_us(request):
...
+                send_slack_notification(
+                    message="A *new contact message* is received. \n *Contact details*: {}"
```
Comment (RishabhJain2018) [inline apps/web/views.py]: A *contact message* is received with details: {}`

File: settings/common.py [hunk 7]
```diff
@@ -274,6 +274,9 @@
 ADMIN_EMAIL = "admin@cloudcv.org"
 CLOUDCV_TEAM_EMAIL = "EvalAI Team <team@cloudcv.org>"
 
+# Slack web hook url
+SLACK_WEB_HOOK_URL = os.environ.get('SLACK_WEB_HOOK_URL', 'http://testslackwebhook.com/webhook')
```
Comment (RishabhJain2018) [inline settings/common.py]: Please add this variable in prod and staging configs.

File: apps/base/utils.py [hunk 8]
```diff
@@ -198,3 +201,23 @@ def get_queue_name(param):
...
+    Keyword Arguments:
+        webhook {string} -- slack webhook URL (default: {settings.SLACK_WEB_HOOK_URL})
+        message {str} -- JSON/Text message to be sent to slack (default: {""})
+    """
+    response = requests.post(
+        webhook,
+        data=json.dumps({'text': str(message)}),
+        headers={'Content-Type': 'application/json'}
+    )
+    if type(response) == requests.models.Response and response.status_code == status.HTTP_200_OK:
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply apps/base/utils.py]: now removed this condition and moved the code in `try/except`, please have a look.

File: apps/base/utils.py [hunk 9]
```diff
@@ -198,3 +201,23 @@ def get_queue_name(param):
...
+        return response
+    else:
+        logger.info(
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply apps/base/utils.py]: done

File: apps/challenges/views.py [hunk 10]
```diff
@@ -39,6 +39,7 @@
 from accounts.serializers import UserDetailsSerializer
 from base.utils import (
     paginated_queryset,
+    send_slack_notification,
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply apps/challenges/views.py]: done

File: apps/challenges/views.py [hunk 11]
```diff
@@ -1123,6 +1124,17 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
 
             zip_config.challenge = challenge
             zip_config.save()
+
+            if settings.SLACK_WEB_HOOK_URL != "http://testslackwebhook.com/webhook":
```
[the remark this answers is not in the document]
Comment (Sanji515) [reply apps/challenges/views.py]: done

File: apps/challenges/views.py (hunk 3 above)
Comment (Sanji515) [reply apps/challenges/views.py]: done

File: apps/challenges/views.py (hunk 2 above)
Comment (Sanji515) [reply apps/challenges/views.py]: done

File: apps/web/views.py (hunk 5 above)
Comment (Sanji515) [reply apps/web/views.py]: done

File: apps/base/utils.py [hunk 12]
```diff
@@ -198,3 +200,31 @@ def get_queue_name(param):
...
+        data = {
+            "text": message["text"],
+            "attachments": [
+                {
+                    "color": "ffaf4b",
+                    "fields": message["fields"]
+                }
+            ]
+        }
+        return requests.post(
```
Comment (RishabhJain2018) [inline apps/base/utils.py]: Why are we returning here? Is it necessary to do that? Can't we simply do `requests.post`

Comment (Sanji515) [reply apps/base/utils.py]: @RishabhJain2018 I'm returning this so that I'm able to test in the `test_utils.py` by checking the returned object type.

Comment (RishabhJain2018) [reply apps/base/utils.py]: okay.

**Pull request merged**